### PR TITLE
Update iced3 to mainline

### PR DESCRIPTION
**Pull request by zapu**
Mainline changed how literals work - instead of having just Literal node, there are now NumberLiterals, StringLiterals, IdentifierLiterals and so forth.

Also, Literals should not be used for object accesses anymore. Mainline introduced PropertyNames for that. So most of iced code that tried to do something like "__it.next()" had to be changed to use `new Access new PropertyName 'next'` instead of `new Access new Value new Literal  'next'`.

All tests pass apart from one test of sourcemaps that also fails in mainline coffee.

This is a merge commit, so it looks kind of messy in the history. I can commit it differently, whatever works best :)

### CODE REVIEW

**Comment by maxtaco** [comment]
👍  LGTM how did the merge go?  Thanks for figuring all of that stuff out!

File: documentation/index.html.js [hunk 1]
```diff
@@ -811,7 +811,7 @@ Expressions
       object that the current function is attached to. If you pass a function as
       a callback or attach it to a different object, the original value of <code>this</code>
       will be lost. If you're not familiar with this behavior,
-      <a href="http://www.digital-web.com/articles/scope_in_javascript/">this Digital Web article</a>
+      <a href="http://64.13.255.16/articles/scope_in_javascript/">this Digital Web article</a>
```
Comment (zapu) [inline documentation/index.html.js:814]: This one kind of bothers me though, the digital-web domain seems to be working correctly. The change was introduced here: https://github.com/jashkenas/coffeescript/commit/f80a8a5120b4b8afd5afecc333486d2306a16303

**Comment by zapu** [comment]
There was some work to update all custom iced code, but other than that everything merged cleanly. 

It's good to have another compiler while working on this, so you are not self-hosting all the time, because every mistake forces a revert. But this gets confusing very quickly (which compiler is invoked right now? which runtime? etc.). It's also good to have a mainline coffee installed alongside, to compare results for some specific cases. 

So while I don't have any systematic workflow, having 3 different nodeenvs with different compilers has "just worked" so far :)

**Comment by zapu** [comment]
So do you want me to rebase/squash all of this to one commit or is it fine as it is?

**Comment by maxtaco** [comment]
I'd say fine as is.  Should I push merge?

**Comment by zapu** [comment]
Give me a minute, I'll run my tests again to be sure it works

**Comment by zapu** [comment]
I did not change code generation in `IcedRuntime` class to use new Literals and PropertyNames but it does not break anything. Maybe I'm misunderstanding usage of IdentifierLiteral and some changes to iced generation were not needed after all.

This pull request installs and works properly, however. So I'd say go ahead with the merge.

**Pull request merged**